src/FlashLoanVault.sol

```diff
@@ -0,0 +1,285 @@
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
+            uint allowedShares = _allowance[owner][receiver] ;
+            require(allowedShares >= shares, "Allowance exceeded!");
+            _allowance[owner][receiver] = allowedShares - shares;
+        }
+
+        //burn wrapped tokens(shares) -> yvDAI 
+        burn(owner, shares);
+        
+        //transfer assets
+        underlying.safeTransfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+    /// @dev Burns shares from owner and sends assets of underlying tokens to receiver; based on the exchange rate.
+    /// @param shares The amount of wrapped tokens to redeem for underlying tokens (assets)
+    /// @param receiver Address of receiver of underlying tokens - DAI
+    /// @param owner Address of owner of Fractional Wrapper shares - yvDAI
+    function redeem(uint256 shares, address receiver, address owner) external returns(uint256 assets) {
+        assets = convertToAssets(shares);
+        
+        // MUST support a redeem flow where the shares are burned from owner directly where owner is msg.sender,
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
+            uint256 allowedShares = _allowance[owner][receiver] ;
+            require(allowedShares >= shares, "Allowance exceeded!");
+            _allowance[owner][receiver] = allowedShares - shares;
+        }
+
+        //burn wrapped tokens(shares) -> yvDAI 
+        burn(owner, shares);
+        
+        //transfer assets
+        underlying.safeTransfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+
+
+    /// @notice Returns the unit 'exchange rate'; assuming 1 unit of underlying was deposited, how much shares would be received 
+    /// @dev wrapperMinted = (underlyingDeposited,(1) * wrapperSupply) / underlyingInWrapper 
+    /// Note: Exchange rate is floating, it's dynamic based on capital in/out-flows
+    /// Note: _totalSupply returns a value extended over decimal precision, in this case 18 dp. hence the scaling before divsion.
+    function getExchangeRate() internal view returns(uint256) {
+        uint256 sharesMinted;
+        return _totalSupply == 0 ? sharesMinted = 1e18 : sharesMinted = (_totalSupply * 1e18) / underlying.balanceOf(address(this));
+    }
+
+    /// @notice Calculate how much yvDAI user should get based on flaoting exchange rate
+    /// @dev getExchangeRate() returns shares minted per unit of underlying asset deposited; at present moment.
+    /// @param assets Amount of underlying tokens (assets) to be converted to wrapped tokens (shares)
+    function convertToShares(uint256 assets) internal view returns(uint256 shares){
+        return assets * getExchangeRate() / 1e18;
```

Comment:
I get what you are doing, but wouldn't this be simpler?
`return _totalSupply == 0 ? 1e18 : assets * _totalSupply / underlying.balanceOf(address(this))`

Reply:
Oh yeah, good point. I did something similar with getExchangeRate() - dunno why it didn't occur to me to carry the same idea forward, lol.

Reply:
Changed it to:

`solidity
    function convertToShares(uint256 assets) internal view returns(uint256){
        return _totalSupply == 0 ? assets : assets * _totalSupply / underlying.balanceOf(address(this));
    }
`

- if _totalSupply is 0, then shares == assets (1:1 peg)
- Else, the conversion is as per formulae above